Add handle_profile_update method and related functionality

alonsoarias opened the pull request:
- Add new handle_profile_update method to process profile field changes
- Implement automatic unenrollment/suspension based on configuration
- Add event triggers for enrollment changes
- Add group membership handling
- Improve cache handling for profile updates
- Add documentation for new functionality

alonsoarias committed:
Add handle_profile_update method and related functionality

- Add new handle_profile_update method to process profile field changes
- Implement automatic unenrollment/suspension based on configuration
- Add event triggers for enrollment changes
- Add group membership handling
- Improve cache handling for profile updates
- Add documentation for new functionality

ndunand commented:
Thanks @alonsoarias for providing this feature! Could you please change the comments within the code to English, so it'll be more easy to read down the line?

I can understand them, but I fear some else in the future might not.

alonsoarias committed:
Add handle_profile_update method and related functionality (Comments translated into English)

- Add new handle_profile_update method to process profile field changes
- Implement automatic unenrollment/suspension based on configuration
- Add event triggers for enrollment changes
- Add group membership handling
- Improve cache handling for profile updates
- Add documentation for new functionality
-

ndunand commented:
Thanks for updating the comments. I tried the updated plugin on my test instante but I can't get to execute the handler, whatever I do.

Can you please provide testing instructions?

alonsoarias committed:
feat: Add attribute expiration behavior tests

Added three new test methods to verify different expiration behaviors:
- testWhenExpiredRemoveBehavior: Tests that users are unenrolled when attributes expire with REMOVE setting
- testWhenExpiredSuspendBehavior: Tests that users are suspended when attributes expire with SUSPEND setting
- testWhenExpiredDoNothingBehavior: Tests that users remain enrolled when attributes expire with DO_NOTHING setting

alonsoarias committed:
Add new event observers for profile field updates

This update ensures that when a user's profile field is modified and no longer  meets the enrollment conditions, the user is either unenrolled or suspended  from the course based on the plugin configuration

alonsoarias commented:
# Testing Instructions for Enrol by User Profile Fields Plugin

## 1. INITIAL SETUP:

```
a) Go to Site administration > Plugins > Enrolments > Enrol by user profile fields
b) Verify the plugin is enabled 
c) In plugin settings:
   - Under "Profile fields to be used in the selector" select at least one custom field
   - Under "Default behaviour after attributes expiration" select desired behavior
```

## 2. CREATE A PROFILE FIELD:

```
a) Go to Site administration > Users > Profile fields
b) Create new field:
   - Shortname: test_field
   - Name: Test Field
   - Data type: Text input
```

## 3. CONFIGURE COURSE ENROLLMENT:

```
a) Go to desired course
b) In Course administration > Users > Enrolment methods
c) Add method "Enrol by user profile fields"
d) In settings:
   - Name: Test enrollment
   - Assign role: Student
   - Behaviour after expiration: Choose behavior (e.g., Suspend user)
   - Rules: Create rule for test_field = "value1"
```

## 4. TEST BEHAVIOR:

```
a) Assign value to user:
   - Edit user profile
   - Set test_field = "value1"
   - User should be automatically enrolled

b) Test expiration:
   - Edit user profile
   - Change test_field to "value2"
   - Run cron or wait for it to execute
   - Verify user is suspended/removed according to configuration
```

## 5. SPECIFIC TEST EXAMPLE:

```bash
# 1. Create profile field "department"
- Shortname: department
- Name: Department
- Type: Text input

# 2. Configure enrollment rule
- Rule: department = "Computer Science"
- When expired: Suspend user
- Role: Student

# 3. Test case 1: Enrollment
- Set user's department = "Computer Science"
- User should be enrolled

# 4. Test case 2: Expiration
- Change user's department = "Mathematics"  
- Wait for cron or run:
  php admin/cli/cron.php
- User should be suspended

# 5. Verify in database:
SELECT u.id, u.username, ue.status 
FROM {user} u
JOIN {user_enrolments} ue ON ue.userid = u.id
WHERE ue.enrolid = [ID_OF_YOUR_ENROL_INSTANCE]
```

## Code Testing Example:

```php
// Enroll user by setting correct field
$user_info_data = new stdClass();
$user_info_data->userid = $user->id;
$user_info_data->fieldid = $field->id;
$user_info_data->data = 'Computer Science';
$DB->insert_record('user_info_data', $user_info_data);

// Force enrollment processing
enrol_attributes_plugin::process_enrolments();

// Verify enrollment
$is_enrolled = is_enrolled(context_course::instance($course->id), $user);
// Should be true

// Test expiration by changing field
$user_info_data->data = 'Mathematics';
$DB->update_record('user_info_data', $user_info_data);

// Force processing
enrol_attributes_plugin::process_enrolments();

// Verify status (if configured to suspend)
$enrolment = $DB->get_record('user_enrolments', ['userid' => $user->id]);
// $enrolment->status should be ENROL_USER_SUSPENDED
```

Please let me know if you need any clarification on any part of these instructions.

ndunand commented:
Thanks @alonsoarias for the detailed testing instructions. With the updated code I now tested succesfully and got the desired behaviour.

ndunand commented:
Hi @alonsoarias ,

Could you have a look at my comment(s) on #57 about groups in particular ? Does the handler really need to call `groups_remove_member()` ?

alonsoarias commented:
Hi @ndunand,
I've analyzed your comments on issue #57 regarding the groups handling. You're right - we shouldn't call groups_remove_member() for all cases. I've submitted a commit that addresses both issues:

For the reactivation problem, I've added the explicit ENROL_USER_ACTIVE parameter to the enrol_user() call, which will properly reactivate suspended users when they meet the criteria again.
For the group membership issue, I've modified the code to only remove users from groups when they are completely unenrolled (ENROL_ATTRIBUTES_WHENEXPIREDREMOVE), while preserving group memberships when they are merely suspended (ENROL_ATTRIBUTES_WHENEXPIREDSUSPEND).

This way, suspension works as a temporary state that preserves group associations, while unenrollment is treated as a more permanent action that removes all course-related memberships.
Let me know if you'd like any further adjustments to the implementation.

ndunand commented:
Thanks @alonsoarias I didn't expect you to react so fast !

In the meantime I've pushed 86bc22e16239fd966a408b015d6bca0265da105f 

But now it's getting too late for me. I'll have a look asap next week.

Kind regards

alonsoarias commented:
No problem at all! I happened to be working on similar issues when I saw your comment. I'm quite interested in this plugin and wanted to contribute to fixing these issues promptly.
I noticed the implementation in 86bc22e and wanted to highlight the key difference between that and my proposed changes (04ed800) for your review next week:
In my implementation, I specifically modified the group handling logic so that:

Users are only removed from groups when completely unenrolled (WHENEXPIREDREMOVE)
Users who are suspended (WHENEXPIREDSUSPEND) remain in their groups

This preserves group memberships during temporary suspension, making the reactivation process more seamless when a user meets the criteria again.
The current implementation in 86bc22e removes users from groups in both cases (suspension and unenrollment), which might not be ideal for temporarily suspended users.
I believe this distinction would provide a better user experience for the reactivation scenario mentioned in issue #57.
Looking forward to your review next week!
Regards,
Alonso
